Onboarding: Allow skipping the welcome page

pfefferle opened the pull request:
The "Welcome" page should be the page, the admin ends up, after the onboarding process.

This PR allows to skip this page, to directly call the `settings` page instead.

<img width="1111" alt="Screenshot 2025-03-26 at 09 13 24" src="https://github.com/user-attachments/assets/7574e8b8-aaef-4784-a09e-689a93fd37df" />

Fixes #

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Re-use the same mechanism that Core uses to dismiss the welcome banner on the dashboard.
* When hidden, the page can be re-activated through the screen-options.
* The tab-menu only shows up if there is more than one tab.


### Other information:

- [ ] Have you written new tests for your changes, if applicable?

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Go to '..'
*

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger a GitHub workflow that will create and push the entry into the branch. -->

<!-- Due to org permissions, the job may fail for PRs crated from a fork under GitHub organizations. -->
<!-- In this case, you can create entry manually with `composer changelog:add` and push it into the branch. -->

<!-- If no changelog entry is required for this PR, please add the "Skip Changelog" label. -->

-   [X] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [X] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [X] Added - for new features
-   [ ] Changed - for changes in existing functionality
-   [ ] Deprecated - for soon-to-be removed features
-   [ ] Removed - for now removed features
-   [ ] Fixed - for any bug fixes
-   [ ] Security - in case of vulnerabilities

#### Message

The option to show/hide the "Welcome Page".

</details>

pfefferle committed:
Onboarding: Allow skipping the welcome page

pfefferle committed:
if $tab is unknown, fall back to to default

pfefferle committed:
phpcs fixes

obenland commented:
Interesting! When I first used the plugin I remember thinking that this would be nice option to have, to just hide the welcome.

Some initial thoughts:
* Maybe we could re-use the same mechanism that Core uses to dismiss notices? It feels to me like a user-level setting.
* Instead of the paragraph and primary button, how about a secondary button or link-button floated to the top right that just says "Hide"?
* When hidden, should we add that information to the hep tab?
* When Welcome is hidden and actor mode doesn't include the blog user, we're left with Settings being the only tab—making tabs obsolete. In that case we shouldn't show them.

pfefferle commented:
thanks for the detailed and early feedback! I agree with all 4 points!

pfefferle committed:
Merge branch 'trunk' into add/onboarding-skip-welcome

pfefferle committed:
Do it welcome screen style

pfefferle committed:
Show screen settings

pfefferle committed:
fix phpcs

pfefferle committed:
Fix PHPCS and save settings

pfefferle committed:
Merge branch 'trunk' into add/onboarding-skip-welcome

pfefferle commented:
@obenland I added all the recommended changes.

* [X] Maybe we could re-use the same mechanism that Core uses to dismiss notices? It feels to me like a user-level setting.
* [X] Instead of the paragraph and primary button, how about a secondary button or link-button floated to the top right that just says "Hide"?

I used the same mechanism as used for the Welcome banner on the dashboard.

<img width="921" alt="Screenshot 2025-03-26 at 09 03 58" src="https://github.com/user-attachments/assets/664a285a-a295-4f95-bbfd-e8d77e488eef" />

* [X] When hidden, should we add that information to the hep tab?

I enabled the "Screen Options" tab and added a checkbox to show/hide the "Welcome Page". I would love to also use that to show/hide the "Advanced Settings": #1449 

<img width="2390" alt="Screenshot 2025-03-26 at 09 04 53" src="https://github.com/user-attachments/assets/e3930faa-5755-477e-933e-655faf12e272" />

* [X] When Welcome is hidden and actor mode doesn't include the blog user, we're left with Settings being the only tab—making tabs obsolete. In that case we shouldn't show them.

<img width="1190" alt="Screenshot 2025-03-26 at 09 02 09" src="https://github.com/user-attachments/assets/63db71b1-a871-4ee7-8bf6-ac82df126018" />

pfefferle committed:
use old text

pfefferle committed:
Merge branch 'add/onboarding-skip-welcome' of https://github.com/Automattic/wordpress-activitypub into add/onboarding-skip-welcome

matticbot committed:
Add changelog

pfefferle committed:
fix typo

pfefferle committed:
Merge branch 'add/onboarding-skip-welcome' of https://github.com/Automattic/wordpress-activitypub into add/onboarding-skip-welcome

obenland commented:
SCREEN OPTIONS!! 😍

obenland commented on includes/wp-admin/class-settings.php:
```diff
@@ -226,25 +241,44 @@ public static function register_settings() {
 				'sanitize_callback' => array( Sanitize::class, 'url_list' ),
 			)
 		);
+
+		\register_setting(
+			'activitypub_welcome',
+			'activitypub_welcome_skip',
+			array(
+				'type'    => 'boolean',
+				'default' => false,
+			)
+		);
```
```suggestion
```
Looks like this is unused now?

obenland commented on includes/wp-admin/class-settings.php:
```diff
@@ -22,6 +22,21 @@ class Settings {
+				}
+
+				return true;
+			},
```
If it's the same to you, let's make that a class method for consistency and so it could be unhooked if need be?

obenland commented on includes/wp-admin/class-settings.php:
```diff
@@ -226,25 +241,44 @@ public static function register_settings() {
 	 * Load settings page.
 	 */
 	public static function settings_page() {
+		$default_tab = \get_user_meta( \get_current_user_id(), 'activitypub_show_welcome_tab', true ) ? 'welcome' : 'settings';
```
There are three calls to get `activitypub_show_welcome_tab` in this method, let's save that in a variable?

obenland commented on includes/wp-admin/class-settings.php:
```diff
@@ -226,25 +241,44 @@ public static function register_settings() {
-		$settings_tabs = array(
-			'welcome'  => array(
+		// Redirect welcome tab to settings if skipped.
+		if ( 'welcome' === $tab && ! \get_user_meta( \get_current_user_id(), 'activitypub_show_welcome_tab', true ) ) {
```
When would this condition be true? When someone manually added `tab=welcome` to URL?

pfefferle replied:
yes

obenland commented on includes/wp-admin/class-settings.php:
```diff
@@ -417,4 +461,49 @@ public static function add_settings_help_tab() {
+		}
+
+		$screen_settings = '<fieldset>
+		<legend class="screen-layout">' . __( 'Settings Pages', 'activitypub' ) . '</legend>
```
```suggestion
		<legend class="screen-layout">' . \esc_html__( 'Settings Pages', 'activitypub' ) . '</legend>
```
This and the following translation functions should be escaped

obenland commented on assets/css/activitypub-admin.css:
```diff
@@ -220,3 +221,24 @@ input.blog-user-identifier {
+	top: 0px;
+	right: 0px;
+	font-size: 13px;
+	padding: 2px 15px 10px 20px;
+	line-height: 1.23076923;
```
```suggestion
	padding: 0 5px 0 20px;
```
How would you feel about this change? It would help with active and focus styles:

Before | After
--- | ---
<img width="834" alt="Screenshot 2025-03-26 at 8 37 38 AM" src="https://github.com/user-attachments/assets/2064bc60-54a8-4d11-9d2c-ab686636afb9" /> | <img width="837" alt="Screenshot 2025-03-26 at 8 37 19 AM" src="https://github.com/user-attachments/assets/b49ae0e4-1455-489e-8a84-e24afa13aaf2" />

obenland commented on includes/wp-admin/class-settings.php:
```diff
@@ -226,25 +241,44 @@ public static function register_settings() {
+			);
+		}
+
+		$settings_tabs['settings'] = array(
+			'label'    => __( 'Settings', 'activitypub' ),
+			'template' => ACTIVITYPUB_PLUGIN_DIR . 'templates/settings.php',
 		);
```
Nitpick, feel free to ignore: we could leave that in the initial `$settings_tabs` declaration

pfefferle replied:
but then the welcome page would be second or we would have to order it afterwards

obenland commented on includes/wp-admin/class-welcome-fields.php:
```diff
@@ -83,6 +83,7 @@ public static function register_welcome_fields() {
 	 */
 	public static function render_welcome_intro_section() {
 		?>
+		<a class="welcome-tab-close" href="<?php echo \esc_url( \admin_url( 'options-general.php?page=activitypub&welcome=0' ) ); ?>" aria-label="Dismiss the welcome page">Dismiss Welcome Page</a>
```
```suggestion
		<a class="welcome-tab-close" href="<?php echo \esc_url( \admin_url( 'options-general.php?page=activitypub&welcome=0' ) ); ?>" aria-label="<?php \esc_attr_e( 'Dismiss the welcome page' ); ?>"><?php \esc_html_e( 'Dismiss Welcome Page' ); ?></a>
```

pfefferle committed:
Update includes/wp-admin/class-settings.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

obenland commented:
Oh, there are some inconsistencies with backslashing global functions, I noticed

pfefferle committed:
Update includes/wp-admin/class-settings.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Update assets/css/activitypub-admin.css

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Update includes/wp-admin/class-welcome-fields.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
make screen_options_show_submit a function

pfefferle committed:
store get_user_meta in a var

pfefferle committed:
globalize some missing functions

pfefferle committed:
Merge branch 'trunk' into add/onboarding-skip-welcome

pfefferle committed:
fix phpcs

pfefferle committed:
Merge branch 'add/onboarding-skip-welcome' of https://github.com/Automattic/wordpress-activitypub into add/onboarding-skip-welcome

pfefferle committed:
Merge branch 'trunk' into add/onboarding-skip-welcome

obenland reviewed (approved):
This turned out pretty nice, kudos!
